sdk/lib/atl/atlimage.h

```diff
@@ -93,26 +93,38 @@ class CImage
 
     HDC GetDC() const throw()
     {
-        if (m_hDC)
-            return m_hDC;
+        ATLASSERT(m_nDCRefCount >= 0);
+
+        if (::InterlockedIncrement(&m_nDCRefCount) == 1)
+        {
+            ATLASSERT(m_hDC == NULL);
+            ATLASSERT(m_hOldBitmap == NULL);
+            m_hDC = ::CreateCompatibleDC(NULL);
+            m_hOldBitmap = (HBITMAP)::SelectObject(m_hDC, m_hBitmap);
+        }
 
-        m_hDC = ::CreateCompatibleDC(NULL);
-        m_hOldBitmap = (HBITMAP)::SelectObject(m_hDC, m_hBitmap);
         return m_hDC;
     }
 
     void ReleaseDC() const throw()
     {
-        ATLASSERT(m_hDC);
+        ATLASSERT(m_nDCRefCount > 0);
 
-        if (m_hDC == NULL)
+        if (::InterlockedDecrement(&m_nDCRefCount) != 0)
+            return;
+
+        if (!m_hDC)
+        {
+            ATLASSERT(!m_hOldBitmap);
             return;
+        }
 
         if (m_hOldBitmap)
```

Comment:
How can `m_hOldBitmap` be `NULL`? An `ASSERT()` would seem more consistent to me.

Reply:
When `SelectObject` failed, it might return `NULL`.

Reply:
Asserts are not compiled in Release mode. So that, in such situation, the GetDC() function from above could fail creating the m_hDC (returning NULL), and then call SelectObject with a NULL hDC, which should (in principle) make that other function fail too (returning a NULL m_hOldBitmap).
If then ReleaseDC() is called later, it will encounter a NULL m_hOldBitmap.

Reply:
> and throw.

_Iff_ MS' ATL CimageDC::GetDC does the same (for behavioural compatibility).